torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,485 @@
+                    param_group_rank = copy.copy(param_group)
+                    param_group_rank["params"] = params
+                    self._partition_parameters[rank].append(param_group_rank)
+
+        return self._partition_parameters
+
+    @property
+    def per_device_params(self) -> Dict[torch.device, List[List[Parameter]]]:
+        """Sorted list of all the params, first per device then per rank.
+
+        Within a list params are sorted per number of elements to allow for an easy bucketing.
+        """
+        if len(self._per_device_params) == 0:
+            # Go through all params, log them per device
+            # The ordering is important here, needs to be the same on all ranks
+            # So that ulterior broadcast calls are matching
+            for param_group in self.param_groups:
+                for param in param_group["params"]:
+                    device = param.device
+                    if self._per_device_params.get(device) is None:
+                        self._per_device_params[device] = [[] for _ in range(self.world_size)]
+                    self._per_device_params[device][self.param_to_rank[param]] += [param]
+
+            # Sort param_lists by size
+            for k in self._per_device_params.keys():
+                for r in self._per_device_params[k]:
+                    r.sort(key=lambda x: x.numel())
+
+        return self._per_device_params
+
+    @property
+    def param_to_rank(self) -> Dict[torch.Tensor, int]:
+        """Look up table to match a given param with a data parallel rank"""
+        if len(self._param_rank) == 0:
+            for rank, param_groups in enumerate(self.partition_parameters()):
+                for param_group in param_groups:
+                    for param in param_group["params"]:
+                        self._param_rank[param] = rank
+        return self._param_rank
+
+    def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) -> Optional[float]:
+        """Performs a single optimization step (parameter update).
+
+        Arguments:
+            closure (callable): A closure that reevaluates the model and
+                returns the loss. Optional for most optimizers.
+        Returns:
+            optional loss, depends on the underlying optimizer
+
+        .. note: Any extra parameter is passed to the base optimizer as-is"""
+
+        # Sync oss param_groups attributes in case they've been updated by a scheduler.
+        self._sync_param_groups()
+
+        # Run the optimizer step on this shard only:
+        if closure is not None:
+            loss = self.optim.step(closure=closure, **kwargs)  # type: ignore
+        else:
+            loss = self.optim.step(**kwargs)
+
+        # Sync all the updated shards in between the ranks
+        self._broadcast_params()
+
+        # Sync hypothethical new results from the wrapped optimizer to the exposed param_groups
+        self._sync_param_groups(local_to_global=True)
```

Comment:
curious, why do we need to sync states like learning rates. Since the owning rank will dictate the updated value of parameters, will it be sufficient to just keep the states on the owner?

Reply:
kind of overkill, but some frameworks are actively using this ([See classy for instance](https://github.com/facebookresearch/ClassyVision/blob/master/classy_vision/optim/classy_optimizer.py#L254)), in both direction (checking optimizer values -> update LR strategy)

Reply:
maybe we can add a followup to make this optional? So that we can avoid the comm overhead for most use cases?

Reply:
ah ok I understand the question a little better now, the name is misleading. It's not actually using the comms, it's syncing in between the `ZeroRedundancyOptimizer.param_groups` (user facing) and `ZeroRedundancyOptimizer.optim.param_groups` (the actual optimizer which only works on a shard). I'll update the name to make that a little clearer